sinks/graphite.py

```diff
@@ -32,77 +43,84 @@ def __init__(self, host="localhost", port=2003, prefix="statsite.", attempts=3,
         if protocol not in ["pickle", "lines"]:
             raise ValueError("Supported protocols are pickle, lines")
 
+        if normalize is not None and normalize not in ("False", "false", "No", "no"):
+            self.normalize_func = self.normalize_key
+        else:
+            self.normalize_func = lambda k: "%s%s" % (self.prefix, k)
+
         self.logger = logging.getLogger("statsite.graphitestore")
         self.host = host
         self.port = port
         self.prefix = prefix
         self.attempts = attempts
         self.sock = self._create_socket()
         self.flush = self.flush_pickle if protocol == "pickle" else self.flush_lines
+        self.metrics = []
 
-    def flush_lines(self, metrics):
+    def normalize_key(self, key):
         """
-        Flushes the metrics provided to Graphite.
+        Take a single key string and return the same string with spaces, slashes and
+        non-alphanumeric characters subbed out and prefixed by self.prefix.
+        """
+        key = SPACES.sub("_", key)
+        key = SLASHES.sub("-", key)
+        key = NON_ALNUM.sub("", key)
+        key = "%s%s" % (self.prefix, key)
+        return key
 
-       :Parameters:
-        - `metrics` : A list of "key|value|timestamp" strings.
+    def append(self, metric):
         """
-        if not metrics:
-            return
+        Add one metric to queue for sending. Addtionally modify key to be compatible with txstatsd
+        format.
 
-        self.logger.info("Outputting %d metrics", len(metrics))
+        :Parameters:
+         - `metric` : A single statsd metric string in the format "key|value|timestamp".
+        """
+        if metric and metric.count("|") == 2:
+            k, v, ts = metric.split("|")
+            k = self.normalize_func(k)
+            self.metrics.append(((k), v, ts))
 
-        # Construct the output, ensure no spaces and slashes in metric name
-        # The data sent must be in the following format:
-        #     <metric path> <metric value> <metric timestamp>
-        # NOTE <metric value> <metric timestamp> are inverted with respect of pickle protocol
-        # http://graphite.readthedocs.io/en/latest/feeding-carbon.html#the-plaintext-protocol
-        lines = list()
-        for m in metrics:
-            if m.count("|") == 2:
-                met = m.split("|")
-                met[0] = met[0].strip().replace(" ", "_").replace("/", "_")
-                lines.append("%s %s %s" % (self.prefix + met[0], met[1], met[2]))
+    def flush_lines(self):
+        """
+        Flushes the metrics provided to Graphite.
+        """
+        if not self.metrics:
+            return
 
+        lines = ["%s %s %s" % metric for metric in self.metrics]
         data = "\n".join(lines) + "\n"
 
         # Serialize writes to the socket
         try:
             self._write_metric(data)
-        except Exception:
+        except StandardError:
```

Comment:
Perhaps an uphill battle which I've lost ;) Previous to python 3 this was the prefered exception to catch in order to avoid accidentally catching `StopIteration`, `GeneratorExit`, `KeyboardInterrupt` and `SystemExit` which are more typically used in flow control vs error handling. Python3 has collapsed the built-in exception inheritance tree to eliminate `StandardError` in favor of `Exception` I /think/ largely in response to people no using StandError.

At any rate, I suppose an argument could be made to do a PR to undo this particular change in favor of writing neutral 2/3 compatible python.